File: ruby/robot-simulator/robot_simulator.rb
class Robot
  DIRECTIONS = %i[north east south west]

  attr_reader :bearing
  attr_accessor :x, :y

  def orient(direction)
    raise ArgumentError unless DIRECTIONS.include?(direction)

    @bearing = direction
  end

  def turn_right
    idx = (DIRECTIONS.index(@bearing) + 1) % DIRECTIONS.size
    @bearing = DIRECTIONS[idx]
  end

  def turn_left
    idx = DIRECTIONS.index(@bearing) - 1
    @bearing = DIRECTIONS[idx]
  end

  def at(x, y)
    @x, @y = x, y
  end

  def coordinates
    [@x, @y]
  end

  def advance
    case @bearing
    when :north
      @y += 1
    when :east
      @x += 1
    when :south
      @y -= 1
    when :west
      @x -= 1
    end
  end
end

class Simulator
  def instructions(str)
    str.each_char.map do |char|
      case char
      when "A" then :advance
      when "R" then :turn_right
      when "L" then :turn_left
      end
    end
  end

  def place(robot, x: 0, y: 0, direction: :north)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, input)
    instructions(input).each { |comm| robot.send(comm) }
  end
end

File: cpp/robot-simulator/robot_simulator.cpp
#include "robot_simulator.h"

using namespace std;

robot_simulator::Robot::Robot(const pair<int, int>& pos, robot_simulator::Bearing b) {
    position = pos;
    bearing = b;

    x = &position.first;
    y = &position.second;
}

pair<int, int> robot_simulator::Robot::get_position() const {
    return position;
}

robot_simulator::Bearing robot_simulator::Robot::get_bearing() const {
    return bearing;
}

void robot_simulator::Robot::turn_right() {
    switch (bearing) {
    case Bearing::NORTH:
        bearing = Bearing::EAST;
        break;
    case Bearing::EAST:
        bearing = Bearing::SOUTH;
        break;
    case Bearing::SOUTH:
        bearing = Bearing::WEST;
        break;
    case Bearing::WEST:
        bearing = Bearing::NORTH;
        break;
    }
}

void robot_simulator::Robot::turn_left() {
    switch (bearing) {
    case Bearing::NORTH:
        bearing = Bearing::WEST;
        break;
    case Bearing::EAST:
        bearing = Bearing::NORTH;
        break;
    case Bearing::SOUTH:
        bearing = Bearing::EAST;
        break;
    case Bearing::WEST:
        bearing = Bearing::SOUTH;
        break;
    }
}

void robot_simulator::Robot::advance() {
    switch (bearing) {
    case Bearing::NORTH: *y += 1; break;
    case Bearing::EAST:  *x += 1; break;
    case Bearing::SOUTH: *y -= 1; break;
    case Bearing::WEST:  *x -= 1; break;
    }
}

void robot_simulator::Robot::execute_sequence(const string& input) {
    for (auto c : input) {
        switch (c) {
        case 'R': turn_right(); break;
        case 'L': turn_left(); break;
        case 'A': advance(); break;
        }
    }
}
